### PR TITLE
SC-3111: Handover of Benchmark Performance Audit Tool in Cloud

### PR DESCRIPTION
**Pull request by zyuzka**
Ticket: https://spryker.atlassian.net/browse/SC-3111

### CODE REVIEW

File: src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+use SprykerSdk\Glue\Benchmark\BenchmarkConfig;
+use SprykerSdk\Shared\Benchmark\Command\AbstractCommandBuilder;
+
+class CommandBuilder extends AbstractCommandBuilder
+{
+    /**
+     * @var \SprykerSdk\Glue\Benchmark\BenchmarkConfig
+     */
+    protected $config;
```
Comment (AsonUnique) [inline src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php]:
```suggestion
    protected $benchmarkConfig;
```

File: src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    protected function getApplication(): string
+    {
+        return 'Glue';
+    }
+
+    /**
+     * @return string
+     */
+    protected function getDefaultTestsDirectory(): string
```
Comment (AsonUnique) [inline src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php]: I guess you can omit `Default` here (or substitute it with `Application`).

File: src/SprykerSdk/Glue/Benchmark/Plugin/Console/BenchmarkRunConsole.php [hunk 3]
```diff
@@ -33,7 +33,9 @@ protected function configure(): void
             ->setDescription(static::COMMAND_DESCRIPTION)
             ->addOption('iterations', null, InputOption::VALUE_OPTIONAL, 'Iterations represent the number of times we will perform the benchmark')
             ->addOption('revs', null, InputOption::VALUE_OPTIONAL, 'The number of times the benchmark is executed consecutively within a single time measurement')
-            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed');
+            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed')
+            ->addOption('report', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed')
```
Comment (AsonUnique) [inline src/SprykerSdk/Glue/Benchmark/Plugin/Console/BenchmarkRunConsole.php]: Please, fix the description of options.

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 4]
```diff
@@ -20,4 +20,34 @@ public function getTestsDirectory(): string
...
     }
+
+    /**
+     * @api
+     *
+     * @return string
+     */
+    public function getReport(): string
+    {
+        return 'generator: "table", cols:["benchmark", "subject", "best", "mean", "worst", "stdev", "revs", "its"]';
```
Comment (AsonUnique) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php]: Usually, default values extracted to class constants.

File: src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    ];
+
+    /**
+     * @param \Generated\Shared\Transfer\PhpBenchConfigurationTransfer $phpBenchConfigurationTransfer
+     *
+     * @return string[]
+     */
+    public function buildCommand(PhpBenchConfigurationTransfer $phpBenchConfigurationTransfer): array
+    {
+        $commandTemplate = self::BASED_PHPBENCH_COMMAND;
```
Comment (AsonUnique) [inline src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php]:
```suggestion
        $commandTemplate = static::BASED_PHPBENCH_COMMAND;
```

File: src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        return sprintf('%s/%s', $this->getFallbackBootstrapFolder(), 'bootstrap.php');
+    }
+
+    /**
+     * @return string
+     */
+    protected function getFallbackBootstrapFolder(): string
+    {
+        $moduleRootFolder = __DIR__ . '/../../../../../..';
```
Comment (AsonUnique) [inline src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php]:
```suggestion
        $moduleRootFolder = dirname(__DIR__,  6);
```

File: src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    /**
+     * @param \Generated\Shared\Transfer\PhpBenchConfigurationTransfer $phpBenchConfigurationTransfer
+     *
+     * @return string
+     */
+    protected function getBootstrapPath(PhpBenchConfigurationTransfer $phpBenchConfigurationTransfer): string
+    {
+        $bootstrapPath = $this->getPathToBootstrap($phpBenchConfigurationTransfer);
+
+        return '--bootstrap=' . $bootstrapPath;
```
Comment (AsonUnique) [inline src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php]: please, consider moving these options to a class constants.

File: src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php (hunk 1 above)
Comment (AsonUnique) [reply src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php]: Please, check Yves and Zed layers too.

File: src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php (hunk 2 above)
Comment (AsonUnique) [reply src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php]: Please, check Yves and Zed layers too.

File: src/SprykerSdk/Yves/Benchmark/Plugin/Console/BenchmarkRunConsole.php [hunk 8]
```diff
@@ -33,7 +33,9 @@ protected function configure(): void
             ->setDescription(static::COMMAND_DESCRIPTION)
             ->addOption('iterations', null, InputOption::VALUE_OPTIONAL, 'Iterations represent the number of times we will perform the benchmark')
             ->addOption('revs', null, InputOption::VALUE_OPTIONAL, 'The number of times the benchmark is executed consecutively within a single time measurement')
-            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed');
+            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed')
```
Comment (AsonUnique) [inline src/SprykerSdk/Yves/Benchmark/Plugin/Console/BenchmarkRunConsole.php]: Please, fix descriptions.

File: src/SprykerSdk/Zed/Benchmark/Communication/Console/BenchmarkRunConsole.php [hunk 9]
```diff
@@ -33,7 +33,9 @@ protected function configure(): void
             ->setDescription(static::COMMAND_DESCRIPTION)
             ->addOption('iterations', null, InputOption::VALUE_OPTIONAL, 'Iterations represent the number of times we will perform the benchmark')
             ->addOption('revs', null, InputOption::VALUE_OPTIONAL, 'The number of times the benchmark is executed consecutively within a single time measurement')
-            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed');
+            ->addOption('path', null, InputOption::VALUE_OPTIONAL, 'Path to the directory that contains tests to be executed')
```
Comment (AsonUnique) [inline src/SprykerSdk/Zed/Benchmark/Communication/Console/BenchmarkRunConsole.php]: Also here

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 10]
```diff
@@ -20,4 +22,34 @@ public function getTestsDirectory(): string
     {
         return APPLICATION_ROOT_DIR . '/tests/Benchmark/';
     }
+
+    /**
+     * @api
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php:32]: Please add specification

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 11]
```diff
@@ -20,4 +22,34 @@ public function getTestsDirectory(): string
...
+     *
+     * @return string
+     */
+    public function getReport(): string
+    {
+        return static::BENCHMARK_REPORT_CONFIG;
+    }
+
+    /**
+     * @api
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php:45]: Please add specification

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 12]
```diff
@@ -20,4 +22,34 @@ public function getTestsDirectory(): string
...
+     *
+     * @return int
+     */
+    public function getIterations(): int
+    {
+        return 1;
+    }
+
+    /**
+     * @api
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php:58]: Please add specification.

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 13]
```diff
@@ -20,4 +22,34 @@ public function getTestsDirectory(): string
...
+     *
+     * @return int
+     */
+    public function getIterations(): int
+    {
+        return 1;
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php]: Maybe it's better to move it to const

File: src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+     * @var \SprykerSdk\Glue\Benchmark\BenchmarkConfig
+     */
+    protected $benchmarkConfig;
+
+    /**
+     * @param \SprykerSdk\Glue\Benchmark\BenchmarkConfig $config
+     */
+    public function __construct(BenchmarkConfig $config)
+    {
+        $this->config = $config;
```
Comment (alexanderM91) [inline src/SprykerSdk/Glue/Benchmark/Command/CommandBuilder.php]: Please rename `$config` to `$benchmarkConfig`

File: src/SprykerSdk/Yves/Benchmark/Command/CommandBuilder.php [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+{
+    /**
+     * @var \SprykerSdk\Yves\Benchmark\BenchmarkConfig
+     */
+    protected $benchmarkConfig;
+
+    /**
+     * @param \SprykerSdk\Yves\Benchmark\BenchmarkConfig $config
+     */
+    public function __construct(BenchmarkConfig $config)
```
Comment (alexanderM91) [inline src/SprykerSdk/Yves/Benchmark/Command/CommandBuilder.php]: Please rename `$config` to `$benchmarkConfig`

File: src/SprykerSdk/Yves/Benchmark/Command/CommandBuilder.php [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    protected function getApplicationTestsDirectory(): string
+    {
+        return sprintf('%s/%s', $this->config->getTestsDirectory(), $this->getApplication());
+    }
+
+    /**
+     * @return int
+     */
+    protected function getIterationConfig(): int
```
Comment (alexanderM91) [inline src/SprykerSdk/Yves/Benchmark/Command/CommandBuilder.php]: Maybe it's better to omit `..Config()`, e.g
It could be `getIteration()`?

Comment (alexanderM91) [reply src/SprykerSdk/Yves/Benchmark/Command/CommandBuilder.php]: Is it an application-specific configuration?

File: src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+
+    /**
+     * @return string
+     */
+    abstract protected function getApplicationTestsDirectory(): string;
+
+    /**
+     * @return int
+     */
+    abstract protected function getIterationConfig(): int;
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/Command/AbstractCommandBuilder.php]: IMO it doesn't make sense for these abstract methods per each application layer.
If I got it correctly we can set the default value inside the console command for Yves, Zed, and Glue.

**Review by alexanderM91 — CHANGES_REQUESTED** [review]

File: src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php [hunk 18]
```diff
@@ -20,4 +24,43 @@ public function getTestsDirectory(): string
...
+     * Specification:
+     * - Returns the configuration to build the Benchmark report.
+     *
+     * @api
+     *
+     * @return string
+     */
+    public function getDefaultReportConfig(): string
+    {
+        return static::BENCHMARK_REPORT_CONFIG;
```
Comment (alexanderM91) [inline src/SprykerSdk/Shared/Benchmark/BenchmarkConfig.php]: An exception will be here, `DEFAULT_BENCHMARK_REPORT_CONFIG`

**Comment by alexanderM91** [comment]
Covered by https://github.com/spryker-sdk/benchmark/pull/3

**Pull request closed without merging**